extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientRecorder.java

```diff
@@ -122,41 +123,53 @@ protected static OidcClient createOidcClient(OidcClientConfig oidcConfig, String
-                        }
-                        MultiMap tokenGrantParams = new MultiMap(io.vertx.core.MultiMap.caseInsensitiveMultiMap());
-
-                        String grantType = oidcConfig.grant.getType() == Grant.Type.CLIENT
-                                ? OidcConstants.CLIENT_CREDENTIALS_GRANT
-                                : OidcConstants.PASSWORD_GRANT;
-                        setGrantClientParams(oidcConfig, tokenGrantParams, grantType);
-
-                        if (oidcConfig.grant.getType() == Grant.Type.PASSWORD) {
-                            Map<String, String> passwordGrantOptions = oidcConfig.getGrantOptions()
-                                    .get(OidcConstants.PASSWORD_GRANT);
-                            tokenGrantParams.add(OidcConstants.PASSWORD_GRANT_USERNAME,
-                                    passwordGrantOptions.get(OidcConstants.PASSWORD_GRANT_USERNAME));
-                            tokenGrantParams.add(OidcConstants.PASSWORD_GRANT_PASSWORD,
-                                    passwordGrantOptions.get(OidcConstants.PASSWORD_GRANT_PASSWORD));
-                        }
-
-                        MultiMap commonRefreshGrantParams = new MultiMap(io.vertx.core.MultiMap.caseInsensitiveMultiMap());
-                        setGrantClientParams(oidcConfig, commonRefreshGrantParams, OidcConstants.REFRESH_TOKEN_GRANT);
-
-                        return new OidcClientImpl(client, tokenRequestUri, grantType, tokenGrantParams,
-                                commonRefreshGrantParams,
-                                oidcConfig);
+                    if (t != null) {
+                        throw toOidcClientException(authServerUri.toString(), t);
+                    }
+
+                    if (tokenRequestUri == null) {
+                        throw new ConfigurationException(
+                                "OpenId Connect Provider token endpoint URL is not configured and can not be discovered");
                     }
+                    MultiMap tokenGrantParams = new MultiMap(io.vertx.core.MultiMap.caseInsensitiveMultiMap());
+
+                    String grantType = (oidcConfig.grant.getType() == Grant.Type.CLIENT
+                            || oidcConfig.grant.getType() == Grant.Type.CLIENT_CREDENTIALS)
+                                    ? OidcConstants.CLIENT_CREDENTIALS_GRANT
+                                    : OidcConstants.PASSWORD_GRANT;
+                    setGrantClientParams(oidcConfig, tokenGrantParams, grantType);
+
+                    // Grant.Type.CLIENT_CREDENTIALS is converted to 'client-credentials' by MP-Config,
+                    // but StringUtil.hyphenate() does not replace '_' by '-', so the expected key is 'client_credentials'.
+                    // Therefore, shall we don't allow Grant.Type.CLIENT_CREDENTIALS but only Grant.Type.CLIENT to not further confuse?
```

Comment:
@sberyozkin, Please let me know your thoughts?

Reply:
@mkowa42 Yes, good point, too much flexibility at the cost of confusing the users is not worth it - I din not realize it this `_` vs `-` issue would arise when asking to support both `client` and `client_credentials` at the same time. In fact now that I think about it, I plan to support `authorization_code` as just `code`.
OK, lets finalize then by using a `client` key only when checking the extra parameters for `client_credentials` grant.

Thanks for your time and patience - look forward to merging this PR soon - take your time please - I'll be a bit delayed tomorrow morning

Cheers

Reply:
@sberyozkin, Changed back to support `client` only.